src/dune_rules/module_compilation.ml

```diff
@@ -49,21 +49,41 @@ let copy_interface ~sctx ~dir ~obj_dir ~cm_kind m =
              (Path.build (Obj_dir.Module.cm_file_exn obj_dir m ~kind:cmi_kind))
            ~dst:(Obj_dir.Module.cm_public_file_exn obj_dir m ~kind:cmi_kind)))
 
-let melange_args (cm_kind : Lib_mode.Cm_kind.t) lib_name module_ =
+let melange_args (cctx : Compilation_context.t) (cm_kind : Lib_mode.Cm_kind.t)
+    module_ =
   match cm_kind with
   | Ocaml (Cmi | Cmo | Cmx) | Melange Cmi -> []
   | Melange Cmj ->
-    let bs_package_name =
-      match lib_name with
-      | None -> []
+    let bs_package_name, bs_package_output =
+      let package_output =
+        Module.file ~ml_kind:Impl module_ |> Option.value_exn |> Path.parent_exn
+      in
+      match Compilation_context.public_lib_name cctx with
+      | None -> ([], package_output)
       | Some lib_name ->
-        [ Command.Args.A "--bs-package-name"; A (Lib_name.to_string lib_name) ]
-    in
-    let package_output =
-      Module.file ~ml_kind:Impl module_ |> Option.value_exn |> Path.parent_exn
+        let dir =
+          let package_output = Path.as_in_build_dir_exn package_output in
+          let lib_root_dir =
+            Path.Build.to_string (Compilation_context.dir cctx)
+          in
+          let src_dir = Path.Build.to_string package_output in
+          let build_context =
```

Comment:
Probably just `(Compilation_context.super_context |> Super_context.context).build_dir` is simpler